### PR TITLE
Feat/event hook

### PR DESCRIPTION
**Pull request by marverlous811**
### Description

Add a feature to send events to 3rd system through hook.

### Related Issue

[Issue 341](https://github.com/8xFF/atm0s-media-server/issues/341)

### Checklist

- [x] I have tested the changes locally.
- [x] I have reviewed the code changes.
- [x] I have updated the documentation if necessary.
- [x] I have added appropriate tests, if applicable.

### CODE REVIEW

**Comment by jibon57** [comment]
@marverlous811 few comments:

- Do you think JW token or some kind of security by header will be better? So that end webhook receiver can verify the request? Simple hash mac with sha256/512 can be calculate with the body & send by header will be better or JW token.
-  Having an unique ID with each request will be helpful as well as. 
- Will system do retry in case HTTP attempt failed first time? In this case request need to go through a queue to make sure requests are sending in sequentially & not breaking it.

**Comment by marverlous811** [comment]
> @marverlous811 few comments:
> 
> * Do you think JW token or some kind of security by header will be better, so that end webhook receiver can verify the request? Simple hash mac with sha256/512 can be calculate with the body & send by header  or JW token.
> * Having an unique ID with each request will be helpful as well as.
> * Will system do retry in case HTTP attempt failed first time? In this case request need to go through a queue to make sure requests are sending in sequentially & not breaking it.

@jibon57:

- I think JWT is a good choice. Because it's the first version, I don't know how the third system calculates our hash. Maybe you can pass your system secret to the connector through the CLI param. I will use it as the key of JWT and the body as the payload, and you can decode and verify it in your system. Do you think that solution is good?
- I missed it. I will update the unique ID for each event tomorrow.
- The first version does not support the feature retry when a message fails. But in the future, we will update to ensure the message is sent successfully by a queue with ACK or maybe a dead-letter queue if need

**Comment by giangndm** [comment]
@marverlous811 @jibon57 I think TLS + Endpoint is enough. For securing environment, we can add some token to endpoint and use TLS for avoiding man-in-middle attack.

**Comment by jibon57** [comment]
@giangndm  most of the time webhook endpoint always remain open/well known. Having a simple token in each request will be very easy to determine whether we need to process the request or not. You can simply calculate hash (md5,sha1 etc.) of the body using secret or add another option to set secret & use a header eg. HASH-SIGNATURE/HASH-TOKEN during sending any request. 

- I think `webhook-url` instated of `hook-uri` will be more appropriate name. 
- @marverlous811 I think retry is an important feature to be have for webhook. You may consider it for sooner.

**Comment by marverlous811** [comment]
@jibon57 I'll update the retry send hook soon, maybe this weekend. 

I use `hook-uri` because we support the webhook and other hook providers such as NATS or Kafka in the future. I use the protocol in URI to detect which node will used. Example: `nats://localhost:4222`. Default is webhook

**Comment by giangndm** [comment]
@marverlous811 new rust version already has async trait, then we don't need async_trait anymore

File: docs/user-guide/features/third-party-system-hook.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+## Message format
+
+Message will sent to another system by using JSON format, defined here:
+
+```typescript
+{
+    node: number,
+    ts: number,
+    event: 'session' | 'peer' | 'remote_track' | 'loacl_track',
```
Comment (giangndm) [inline docs/user-guide/features/third-party-system-hook.md:28]: have typo issue

File: packages/media_connector/src/hook_producer.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            }),
+            peer_event::Event::Stats(_params) => None,
+        };
+        if let Some(hook_data) = hook_data {
+            self.queue.push_back(hook_data);
+        }
+        Some(())
+    }
+
+    pub async fn on_tick(&mut self) -> Option<()> {
```
Comment (giangndm) [inline packages/media_connector/src/hook_producer.rs:178]: We need to send more than one event per tick, the queue will full very soon

**Review by giangndm — CHANGES_REQUESTED** [review]

**Comment by marverlous811** [comment]
@giangndm, I have written new code that uses workers to send the hook. The Github action build job reported that our version of SQLX needs to be updated. Do you think we should update it now?

**Comment by giangndm** [comment]
@marverlous811 sea-orm crate has some outdate dependence, may be we need to create an issue in sea-orm repo

**Comment by giangndm** [comment]
@marverlous811 I just updated for fixing cargo-deny issue, please merge with latest source code

**Comment by giangndm** [comment]
@marverlous811 I just refactored the PR. For simplicity, I used a direct struct generated from protobuf and converted it to binary or JSON based on the configuration parameter.

@jibon57 I agree that we should have some form of data protection. In the current implementation, to avoid making it more complicated, we can use a simple solution by implementing a hook endpoint that includes secure query or path parameters. With a HASH-SIGNATURE, we can place it into the headers, allowing us to implement it in the future without any BREAKING-CHANGES.

**Comment by jibon57** [comment]
@giangndm thanks. Yes, simple hash calculated by using body + secret and placed in header during sending post request to the webhook end point URL will be enough for the moment. This way the webhook end point can verify either it was coming from `atm0s-media-server` or not. Here we can use default secret or can have another option to set secret specifically for webhook.

**Comment by giangndm** [comment]
@marverlous811 @jibon57 I merged it as a first version. Security or retry-backoff will be implemented soon

**Pull request merged**